### PR TITLE
Emit thunk function for specific ABI on WebAssembly.

### PR DESCRIPTION
**Pull request by kateinoigakukun**
Changed to make thunk to convert thin-to-thick and non-throws-to-throws. 
We needs it on WebAssembly host because WASM checks number of arguments strictly for indirect function call.

This patch allows you to run the Swift code below.

```swift
func f(_ a: (Int) -> Void) {
    g(a)
}

func g(_ b: (Int) throws -> Void) {
    try! b(1)
}

f { _ in }
```

### CODE REVIEW

**Comment by zhuowei** [comment]
Wow! This is amazing! Thank you so much!

I haven't tested this: does this also handle the [optional protocol witness parameters](https://github.com/apple/swift/blob/9f5bf22ee3a8cbfe4c090c6c9f2652f5a1f726dd/lib/IRGen/GenFunc.cpp#L38)?

**Comment by kateinoigakukun** [comment]
Thanks 😊

> does this also handle the optional protocol witness parameters?


I didn't care about that and I can't figure out how the problem happens. Could you give me an example for the case?

**Comment by kateinoigakukun** [comment]
My branch was broken by rebasing upstream swiftwasm branch 😅

**Comment by MaxDesiatov** [comment]
I'm sorry about that, I thought about using merge commits instead of rebase, but downside is that the main swift repository uses merge commits too. Adding our own would potentially make commit history quite unreadable. Please let us know if you think there's a better solution 🙏

**Comment by kateinoigakukun** [comment]
@MaxDesiatov Thanks for maintaining this downstream repo 👍 
I have no idea to solve this problem. All we can do is "moving fast". 😅

**Comment by kateinoigakukun** [comment]
@zhuowei Sorry for pinging you. Can I merge this patch?

File: lib/SIL/TypeLowering.cpp [hunk 1]
```diff
@@ -2538,12 +2538,25 @@ TypeConverter::checkFunctionForABIDifferences(SILFunctionType *fnTy1,
...
+      return ABIDifference::NeedsThunk;
+    }
+  }
+
   auto rep1 = fnTy1->getRepresentation(), rep2 = fnTy2->getRepresentation();
   if (rep1 != rep2) {
     if (rep1 == SILFunctionTypeRepresentation::Thin &&
-        rep2 == SILFunctionTypeRepresentation::Thick)
+        rep2 == SILFunctionTypeRepresentation::Thick) {
+      // There is no ABI compatibility between thin and think on WebAssembly,
```
Comment (MaxDesiatov) [inline lib/SIL/TypeLowering.cpp]: typo: "think and thick"

Comment (kateinoigakukun) [reply lib/SIL/TypeLowering.cpp]: oops!

**Comment by MaxDesiatov** [comment]
When building an SDK with this branch using [swiftwasm-package-sdk](https://github.com/swiftwasm/swiftwasm-package-sdk/) and using the packaged `swiftwasm` script I get this error:

```
~/swiftwasm-package-sdk/output/swiftwasm-sdk$ ./swiftwasm example/hello.swift out.wasm
wasm-ld: error: ./opt/swiftwasm-sdk/lib/swift_static/wasm/wasm32/swiftrt.o: Bad section type
```

I currently don't have any knowledge of what this file is expected to contain, a basic check shows this:

```
~/swiftwasm-package-sdk/output/swiftwasm-sdk$ file opt/swiftwasm-sdk/lib/swift_static/wasm/wasm32/swiftrt.o 
opt/swiftwasm-sdk/lib/swift_static/wasm/wasm32/swiftrt.o: WebAssembly (wasm) binary module version 0x1 (MVP)
```

I've built everything using this branch with `swiftwasm` branch merged in and its `build-linux.sh` script, then cloned [swiftwasm-package-sdk](https://github.com/swiftwasm/swiftwasm-package-sdk/), put the compiled prebuilts and used `./buildPackages.sh` to get the package.

@kateinoigakukun hope you could provide some pointers to where things could have gone wrong?

**Comment by MaxDesiatov** [comment]
My current hunch is that this broke after merging #11, maybe a simple update to the [`swiftwasm`](https://github.com/swiftwasm/swiftwasm-package-sdk/blob/master/sdkroot/swiftwasm) script could fix this? Should `"$sdk/extra_objs/swift_start.o"` and `"$sdk/extra_objs/swift_end.o"` be removed from the list of linker arguments?

**Comment by MaxDesiatov** [comment]
Just got the artifacts uploaded for `swiftwasm-sdk` and updated links for those in `swiftwasm-package-sdk`, now this is stably reproducible this way (these commands are for macOS, but I got the same message on Linux as described a few comments above):

```
git clone https://github.com/swiftwasm/swiftwasm-package-sdk.git
cd swiftwasm-package-sdk
./downloadPrebuilts.sh
./buildPackages.sh
cd output
tar xf swiftwasm-sdk-macos.tar.xz
cd swiftwasm-sdk
./swiftwasm example/hello.swift hello.wasm
```

yields 
```
wasm-ld: error: ./opt/swiftwasm-sdk/lib/swift_static/wasm/wasm32/swiftrt.o: Bad section type
```

It's not that the isue is caused by this PR, but it seems to be caused by #11 or some other changes since the last binary packages were built, and it currently prevents me from testing this PR.

**Comment by kateinoigakukun** [comment]
@MaxDesiatov I think `Bad section type` error is due to your old `wasm-ld`. (I faced same error after rebased upstream `apple/swift`.)
Please use `wasm-ld` 9.0.0 or above.

```sh
$ wasm-ld --version
LLD 9.0.0 (https://github.com/llvm/llvm-project 0399d5a9682b3cef71c653373e38890c63c4c365)
```

If you can't resolve by updating `wasm-ld`, please try to validate the object file using [`wasm-validate`](https://github.com/WebAssembly/wabt)

**Comment by MaxDesiatov** [comment]
@kateinoigakukun thanks for the tip!

I've managed to get fresh `wasm-ld` and the bad section error went away, but now I stumble upon these new errors both with `build.sh` script from your `swiftwasm-scripts` project and `swiftwasm` script from our `swiftwasm-package-sdk`:

```
wasm-ld: error: ./opt/swiftwasm-sdk/lib/swift_static/wasm/libswiftCore.a: archive has no index; run ranlib to add one
wasm-ld: error: ./opt/swiftwasm-sdk/lib/swift_static/wasm/libswiftImageInspectionShared.a: archive has no index; run ranlib to add one
```

**Comment by kateinoigakukun** [comment]
@MaxDesiatov Ah, I faced same error too `archive has no index; run ranlib to add one`. You may be  using `ranlib` and `ar` built in Xcode. Please ensure that `CMAKE_AR` and `CMAKE_RANLIB` points `llvm-ar` and `lllvm-ranlib` in CMakeCache.txt

**Comment by MaxDesiatov** [comment]
This PR will need to be rebased and force-pushed on top of `swiftwasm` at some point, I had to rebase `swiftwasm` on top of `master` in an attempt to resolve compatibility issues with the latest `llvm-project`, which in turn was apparently needed for latest `wasi-sdk`.

**Comment by MaxDesiatov** [comment]
Seems legit, let's get it merged 🙌

**Comment by hggz** [comment]
https://forums.swift.org/t/reality-vs-reality/31835/24

HELP OPEN THE FLOOD GATES

**Pull request merged**